pkg/kubelet/pod_workers.go

```diff
@@ -221,76 +602,412 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
 			p.managePodLoop(podUpdates)
 		}()
 	}
-	if !p.isWorking[pod.UID] {
-		p.isWorking[pod.UID] = true
-		podUpdates <- *options
-	} else {
-		// if a request to kill a pod is pending, we do not let anything overwrite that request.
-		update, found := p.lastUndeliveredWorkUpdate[pod.UID]
-		if !found || update.UpdateType != kubetypes.SyncPodKill {
-			p.lastUndeliveredWorkUpdate[pod.UID] = *options
+
+	// dispatch a request to the pod worker if none are running
+	if !status.IsWorking() {
+		status.working = true
+		podUpdates <- work
+		return
+	}
+
+	// never lose a PodStatusFunc, but invoke them in order
+	if undelivered, ok := p.lastUndeliveredWorkUpdate[pod.UID]; ok {
+		// track the max latency between when a config change is requested and when it is realized
+		// NOTE: this undercounts the latency when multiple requests are queued, but captures max latency
+		if !undelivered.Options.StartTime.IsZero() && undelivered.Options.StartTime.Before(work.Options.StartTime) {
+			work.Options.StartTime = undelivered.Options.StartTime
+		}
+
+		if originalKillOpts := undelivered.Options.KillPodOptions; originalKillOpts != nil {
+			// once a pod is killed, all subsequent UpdatePod methods must be "kill" and so we should
+			// be able to carry the PodStatusFunc over
+			if originalKillOpts.PodStatusFunc != nil {
```

Comment:
Do we need to accumulate `PodStatusFunc`? There is only one pod status, and later one will overwrite the previous one? In this case, can we just keep the latest one?

I'm still not quite comfortable with adding this recursive function if not needed.

If the caller wants to perform operations with side effect in the function, that is a misusage, they should probably use the `CompleteCh` instead in that case.

Reply:
So in theory you could evict a pod and then the user could delete it faster (so you'd lose the eviction state) by reducing the grace period.  That's the primary bug this fixes.

You could also delete a pod and then have it evicted, in which case we need both data to pass through.  You could also be preempted then evicted - what status should be on the pod?

Reply:
Most of the functions are just updating `Phase`, `Message` and `Reason`, and the updates are applied to the `status` passed in this function, not to the status manager or apiserver, so the change history is discarded directly.

In this case, what status has gone through doesn't actually matter, only the latest status matters.
* If newer status doesn't completely overwrite the old status, the generated mixed status may be confusing anyway;
* If newer status does completely overwrite the old status, the old status is not useful.

This seems only to be meaningful to me if we update the status manager in between.